### PR TITLE
Weekly reporting

### LINKED ISSUE #4: Weekly Reports for customers
- Weekly hours report for teams to handle (i.e approve or share or etc). 
- Slack integration to remind them. 
- Slack integration (company-wide recurring) for the status of handled reports (what reports have been handled and what not. )

### PR DESCRIPTION
**Pull request by jale2014**
Fixes #4

### CODE REVIEW

**Comment by ojarva** [comment]
That is a nice small PR with very good commit messages :trollface:

File: invoices/views.py [hunk 1]
```diff
@@ -279,6 +308,8 @@ def queue_update(request):
 def get_pdf(request, invoice_id, pdf_type):
     if pdf_type == "hours":
         pdf, title = generate_hours_pdf_for_invoice(request, invoice_id)
+    elif pdf_type == "weekly":
+        pdf, title = generate_hours_pdf_for_weekly_report(request, invoice_id) #invoice_id is actually weekly_report_id
```
Comment (ojarva) [inline invoices/views.py]: > invoice_id is actually weekly_report_id

This is not a proper way to do this - if the argument name says it is `invoice_id`, it won't be something else.

File: invoices/utils.py [hunk 2]
```diff
@@ -173,10 +187,36 @@ def match_invoice(self, data):
...
+        weekly_report = self.weekly_report_data.get(weekly_report_key)
+        if weekly_report is not None:
+            logger.debug("Weekly report already exists: %s", weekly_report_key)
+            if weekly_report.tags != data["project_tags"]:
+                weekly_report.tags = data["project_tags"]
+                weekly_report.save()
+            return weekly_report
+        else:
+            logger.info("Creating a new weekly report %s", weekly_report_key)
+            weekly_report, _ = WeeklyReport.objects.update_or_create(
```
Comment (ojarva) [inline invoices/utils.py]: If you are always creating, don't use `update_or_create`, use `weekly_report = WeeklyReport(...)\ weekly_report.save()` instead.

File: invoices/utils.py [hunk 3]
```diff
@@ -136,18 +144,24 @@ def get_invoices():
         invoices_data[invoice_key] = invoice
     return invoices_data
 
+def get_weekly_reports():
+    weekly_report_data = {}
+    for weekly_report in WeeklyReport.objects.all():
```
Comment (ojarva) [inline invoices/utils.py]: Wait, why are you fetching all weekly reports, instead of what is actually necessary?

File: invoices/utils.py [hunk 4]
```diff
@@ -277,3 +318,41 @@ def refresh_stats(start_date, end_date):
...
+                    end_date,
+                    weekly_reports.count())
+    else:
+        logger.info("Updating statistics for all weekly reports")
+        weekly_reports = WeeklyReport.objects.all()
+    for weekly_report in weekly_reports:
+        entries = HourEntry.objects.filter(weekly_report=weekly_report).filter(incurred_hours__gt=0)
+        aws_entries = None
+        stats = calculate_weekly_entry_stats(entries, aws_entries)
+        for field in STATS_FIELDS:
```
Comment (ojarva) [inline invoices/utils.py]: The rest of the method seems to be exact copy of what is in `refresh_stats`, so splitting that to a well-named method would make sense, no?

File: invoices/views.py [hunk 5]
```diff
@@ -587,8 +635,11 @@ def invoice_page(request, invoice_id, **_):
 
 
     previous_invoices = []
+    recent_weekly_report = None
     if invoice.project_m:
         previous_invoices = Invoice.objects.filter(project_m=invoice.project_m)
+        weekly_reports = WeeklyReport.objects.filter(project_m=invoice.project_m, client=invoice.client)
```
Comment (ojarva) [inline invoices/views.py]: `Project` model always identifies client as well, so it should not be necessary to filter by `client` as well

File: invoices/views.py [hunk 6]
```diff
@@ -587,8 +635,11 @@ def invoice_page(request, invoice_id, **_):
...
+        recent_weekly_report = weekly_reports[0]
```
Comment (ojarva) [inline invoices/views.py]: What happens here if there is no weekly reports for some reason? Does `invoice.project_m` guarantee there is always weekly reports available?

**Comment by ojarva** [comment]
To make this simpler, do not create any weekly reports without `project_m` set. With that, there is no reason to have `client`/`project`/... fields in `WeeklyReport` model, as those are already available in `Project` model.
The whole "stuff with no projects" is kind of legacy from the period when the system was using official 10kft APIs with more data inconsistencies. That should be removed altogether in the future (not in this PR, though).

**Comment by ojarva** [comment]
Please install python3 `pep8` and run `pep8 --ignore=E501,E402,E701` for all the code you have added, and fix all the relevant errors, such as missing spaces, extra spaces, missing line breaks etc.

File: invoices/invoice_utils.py [hunk 7]
```diff
@@ -147,8 +147,13 @@ def combine_invoice_parts(*combine_stats):
                 stats[k].update(v)
     return stats
 
-def calculate_entry_stats(hour_entries, fixed_invoice_rows, aws_entries=None):
+def calculate_entry_stats(hour_entries, fixed_invoice_rows=None, aws_entries=None):
```
Comment (ojarva) [inline invoices/invoice_utils.py]: What is the reason for adding default value (`None`) here?

File: invoices/templates/weekly_report_hours.html [hunk 8]
```diff
@@ -0,0 +1,47 @@
+{% extends "base.html" %}
```
Comment (ojarva) [inline invoices/templates/weekly_report_hours.html:1]: These seem to be almost exact copies of per-invoice report pages. If that is the case, please split common part to a separate file and include it.

File: invoices/templates/snippets/weekly_report.html [hunk 9]
```diff
@@ -0,0 +1,125 @@
+{% load humanize %}
```
Comment (ojarva) [inline invoices/templates/snippets/weekly_report.html:1]: These seem to be almost exact copies of per-invoice report pages. If that is the case, please split common part to a separate file and include it.

File: invoices/templates/snippets/last_week_report.html [hunk 10]
```diff
@@ -0,0 +1,37 @@
+<h2>Changes since last week</h2>
```
Comment (ojarva) [inline invoices/templates/snippets/last_week_report.html]: These seem to be almost exact copies of per-invoice report pages. If that is the case, please split common part to a separate file and include it.

File: invoices/templates/personweek.html [hunk 11]
```diff
@@ -0,0 +1,79 @@
+{% extends "base.html" %}
```
Comment (ojarva) [inline invoices/templates/personweek.html]: These seem to be almost exact copies of per-invoice report pages. If that is the case, please split common part to a separate file and include it.

File: invoices/pdf_utils.py [hunk 12]
```diff
@@ -45,3 +46,27 @@ def generate_hours_pdf_for_invoice(request, invoice):
...
+                                  weekly_report_data.project,
+                                  weekly_report_data.year,
+                                  weekly_report_data.week)
+    title = title.replace(u"\xe4", u"a").replace(u"\xb6", u"o").replace(u"\x84", u"A").replace(u"\x96", u"O").replace(u"\xf6", "o")
+
+    entries = HourEntry.objects.filter(project=weekly_report_data.project,
+                                       client=weekly_report_data.client,
+                                       date__year__gte=weekly_report_data.year,
+                                       date__week=weekly_report_data.week).filter(incurred_hours__gt=0)
+    phases = {}
```
Comment (ojarva) [inline invoices/pdf_utils.py]: This again is copy-pasted. Do not copy-paste (major) parts when that can be easily avoided.

File: invoices/models.py [hunk 13]
```diff
@@ -187,6 +188,106 @@ class Meta:
...
+        self.weekly_report_state = "C"
+        if comment.checked:
+            self.weekly_report_state = "A"
+        return self.weekly_report_state
+
+    def get_tags(self):
+        if self.tags:
+            return self.tags.split(",")
+
+    def compare(self, other):
```
Comment (ojarva) [inline invoices/models.py]: Is this whole `compare` exact duplicate of what is in `Invoice`? Is it going to stay that for foreseeable future? If the answer to both is yes, you know what needs to be done.

File: invoices/utils.py (hunk 3 above)
Comment (jale2014) [reply invoices/utils.py]: This works similar to get_invoices(). Shouldn't this work the same way, i.e. fetch all weekly reports from DB? What do you mean by "what is actually necessary"?

File: invoices/utils.py (hunk 2 above)
Comment (jale2014) [reply invoices/utils.py]: This works similar to "match_invoice". Should it be changed too?

File: invoices/utils.py (hunk 3 above)
Comment (ojarva) [reply invoices/utils.py]: You're right, `get_invoices` has the same thing. There is no need to fetch all the invoices/weekly reports when updating a specific time range, but maybe it is so much simpler it should stay that way.

File: invoices/utils.py (hunk 2 above)
Comment (ojarva) [reply invoices/utils.py]: I think yes - there is something wrong if the previous step did not find the report (/invoice), and it is created in this one. Also, a tiny detail in that `if` statement - there is no need to add `is not None`, writing just `if weekly_report` is a typical python idiom for checking it.

File: invoices/templates/snippets/weekly_report.html (hunk 9 above)
Comment (jale2014) [reply invoices/templates/snippets/weekly_report.html:1]: The common part is now included in "snippets/project_hours.html"

File: invoices/templates/weekly_report_hours.html (hunk 8 above)
Comment (jale2014) [reply invoices/templates/weekly_report_hours.html:1]: The common part is now included in "snippets/report_hours.html"

**Comment by jale2014** [comment]
All code review comments should be addressed in the review fixes

File: invoices/models.py [hunk 14]
```diff
@@ -241,7 +216,7 @@ def date(self):
 
     @property
     def full_name(self):
```
Comment (ojarva) [inline invoices/models.py]: Should full name include week and year as well, or is skipping those intentional?

Comment (jale2014) [reply invoices/models.py]: Yes it is intentional, weekly_report_page.html displays both full_name and date (incl. year & week). This is similar to what is done in invoice_page.html

**Pull request merged**